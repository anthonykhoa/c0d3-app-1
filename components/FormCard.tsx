import React from 'react'
import { Button } from './theme/Button'
import { MdInput } from './MdInput'

// capitalizes first letter of a string
const capitalizeFirst = (str: string) => {
  return str.replace(/./, char => char.toUpperCase())
}

export const TEXT_AREA = 'TEXT_AREA'

export const MD_INPUT = 'MD_INPUT'

export type Option = {
  title: string
  type?: string | undefined
  error?: string
  placeHolder?: string
  value?: string | number
}

type Btn = {
  title?: string
  onClick: (value: any) => void
}

type FormCardProps = {
  values: Option[]
  onSubmit: Btn
  capitalizeTitle?: boolean
  onChange: Function
  title?: string
}

type OptionsListProps = {
  onChange: Function
  capitalizeTitle: boolean
  options: Option[]
}

const titleStyle: React.CSSProperties | undefined = {
  textDecoration: 'underline',
  marginBottom: 20
}

const formStyle: React.CSSProperties | undefined = {
  textAlign: 'center',
  padding: '10px',
  border: '2px solid rgb(0,0,0,.3)',
  backgroundColor: 'white'
}

const optionStyle: React.CSSProperties | undefined = {
  border: '3px solid rgb(84, 64, 216, .3)',
  padding: '10px',
  backgroundColor: 'rgb(84, 64, 216, .04)',
  textAlign: 'left'
}

export const OptionsList: React.FC<OptionsListProps> = ({
  onChange,
  capitalizeTitle,
  options
}) => {
  const inputs = options.map((option: Option, i: number) => {
    const { title, value, placeHolder, type } = option
    if (option.title === 'id') return []
    const inputType =
      type === MD_INPUT ? (
        <MdInput
          bgColor="white"
          value={`${value || ''}`}
          onChange={(value: string) => {
            onChange(value, i)
          }}
        />
      ) : (
        <input
          data-testid={`input${i}`}
          style={{ border: '1px solid rgb(84, 64, 216, .3)' }}
          type="text"
          value={`${value || ''}`}
          onChange={e => {
            onChange(e.target.value, i)
          }}
          placeholder={placeHolder || ''}
        />
      )
    return (
      <div
        key={i}
        className="d-flex flex-column"
        style={{
          ...optionStyle,
          marginBottom: i === options.length - 1 ? 0 : 10
        }}
      >
        <h5 data-testid={`h5${title}${i}`}>
          {(capitalizeTitle && capitalizeFirst(title)) || title}
        </h5>
        {inputType}
        {option.error && (
          <>
            <h6 className="text-danger">{option.error}</h6>
          </>
        )}
      </div>
    )
  })

  return <>{inputs}</>
}

export const FormCard: React.FC<FormCardProps> = ({
  values,
  onSubmit,
  capitalizeTitle = true,
  title,
  onChange
}) => {
  const btnOnClick = () => onSubmit.onClick(values)

  return (
    <div style={formStyle} className="rounded">
      {title && <h2 style={titleStyle}>{title}</h2>}
      <OptionsList
        capitalizeTitle={capitalizeTitle}
        options={values}
        onChange={onChange}
      />
      <div style={{ margin: 10 }}>
        <Button onClick={btnOnClick} type="success">
          {onSubmit.title}
        </Button>
      </div>
    </div>
  )
}
