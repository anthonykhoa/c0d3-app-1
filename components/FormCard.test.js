import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { FormCard } from './FormCard'

let testBtnOnClick = ''

const mockBtn = {
  title: 'Create New Challenge',
  onClick: ok => (testBtnOnClick = ok)
}

let mockValues = [{ title: 'super' }]

describe('FormCard Component', () => {
  test('Input value should be nothing when value prop is not there', () => {
    const { container, queryByTestId } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const input = queryByTestId('inputsuper')

    expect(input.value).toEqual('')
    expect(container).toMatchSnapshot()
  })

  test('Should render text onto textbox when user types', () => {
    const { container, queryByTestId } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const input = queryByTestId('inputsuper')

    fireEvent.change(input, {
      target: { value: 'Javascript' }
    })

    expect(input.value).toContain('Javascript')
    expect(container).toMatchSnapshot()
  })

  test('Should render MdInput component when type is MD_INPUT', () => {
    mockValues = [{ title: 'fireStarter', type: 'MD_INPUT' }]
    const { container, queryByTestId } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const input = queryByTestId('textbox')
    fireEvent.change(input, {
      target: { value: 'Javascript' }
    })

    expect(input.value).toContain('Javascript')
    expect(container).toMatchSnapshot()
  })

  test('Should call function with updated array as parameter when button is clicked', () => {
    const { container, queryByText } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const button = queryByText('Create New Challenge')
    fireEvent.click(button, { button: 1 })
    expect(testBtnOnClick).toEqual([
      {
        title: 'fireStarter',
        type: 'MD_INPUT',
        value: 'Javascript'
      }
    ])
    expect(container).toMatchSnapshot()
  })

  test('Should not call onSubmit callback function if there is a nums error', () => {
    testBtnOnClick = 'cat'
    mockValues = [...mockValues, { title: 'fiary', error: ['nums'] }]
    const { container, queryByText } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const button = queryByText('Create New Challenge')
    fireEvent.click(button, { button: 1 })
    expect(testBtnOnClick).toEqual('cat')
    expect(container).toMatchSnapshot()
  })

  test('Should not call onSubmit callback function if there is a require error', () => {
    mockValues = [...mockValues, { title: 'fiary', error: ['require'] }]
    const { container, queryByText } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const button = queryByText('Create New Challenge')
    fireEvent.click(button, { button: 1 })
    expect(testBtnOnClick).toEqual('cat')
    expect(container).toMatchSnapshot()
  })

  test('Should output empty title when title is empty', () => {
    mockValues = [{ title: '' }]
    const { container, getByTestId } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    const input = getByTestId('h50')
    expect(input.value).toBeFalsy
    expect(container).toMatchSnapshot()
  })

  test('Should render nothing if id is a title', () => {
    mockValues = [{ title: 'id' }]
    const { container } = render(
      <FormCard values={mockValues} onSubmit={mockBtn} />
    )
    expect(container).toBeFalsy
    expect(container).toMatchSnapshot()
  })
})
