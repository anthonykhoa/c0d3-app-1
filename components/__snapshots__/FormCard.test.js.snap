// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCard Component Should call function with updated array as parameter when button is clicked 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 0px;"
    >
      <h5
        data-testid="h5fireStarter0"
      >
        FireStarter
      </h5>
      <div
        style="background-color: white;"
      >
        <button
          class="btn"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </button>
        <button
          class="btn"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </button>
        <textarea
          data-testid="textbox"
          placeholder="Type something..."
          style="padding: 1rem; width: 100%;"
        />
      </div>
    </div>
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`FormCard Component Should output empty title when title is empty 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 0px;"
    >
      <h5
        data-testid="h50"
      />
      <input
        data-testid="input0"
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`FormCard Component Should render MdInput component when type is MD_INPUT 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 0px;"
    >
      <h5
        data-testid="h5fireStarter0"
      >
        FireStarter
      </h5>
      <div
        style="background-color: white;"
      >
        <button
          class="btn"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </button>
        <button
          class="btn"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </button>
        <textarea
          data-testid="textbox"
          placeholder="Type something..."
          style="padding: 1rem; width: 100%;"
        >
          Javascript
        </textarea>
      </div>
    </div>
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`FormCard Component Should render nothing if id is a title 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`FormCard Component Should update testing onChange variable when person types 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 0px;"
    >
      <h5
        data-testid="h5fireStarter0"
      >
        FireStarter
      </h5>
      <input
        data-testid="input0"
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`FormCard Component Should update testing onChange variable when person types MdInput 1`] = `
<div>
  <div
    class="rounded"
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 0px;"
    >
      <h5
        data-testid="h5fireStarter0"
      >
        FireStarter
      </h5>
      <div
        style="background-color: white;"
      >
        <button
          class="btn"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </button>
        <button
          class="btn"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </button>
        <textarea
          data-testid="textbox"
          placeholder="Type something..."
          style="padding: 1rem; width: 100%;"
        >
          Javascript
        </textarea>
      </div>
    </div>
    <div
      style="margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;
